common/metagraph_syncer.py

```diff
@@ -0,0 +1,145 @@
+import asyncio
+from concurrent.futures import ThreadPoolExecutor
+import dataclasses
+from dataclasses import field
+from datetime import datetime
+import functools
+import bittensor as bt
+from typing import Dict, List, Callable, Optional
+import threading
+import traceback
+
+from common import utils
+
+
+class MetagraphSyncer:
+    @dataclasses.dataclass
+    class _State:
+        metagraph: Optional[bt.metagraph] = None
+        last_synced_time: Optional[datetime] = None
+        listeners: List = field(default_factory=list)
+
+    def __init__(self, subtensor: bt.subtensor, config: Dict[int, int]):
+        """Constructs a new MetagraphSyncer, that periodically refreshes metagraph defined in the config.
+
+        Args:
+            subtensor (bt.subtensor): The subtensor used to fetch the metagraphs.
+            config (Dict[int, int]): A mapping of netuid to the cadence (in seconds) to sync the metagraph.
+        """
+        self.subtensor = subtensor
+        self.config = config
+        self.metagraph_map: Dict[int, MetagraphSyncer._State] = {
+            netuid: MetagraphSyncer._State() for netuid in config.keys()
+        }
+        self.is_running = False
+        self.done_initial_sync = False
+        self.lock = threading.RLock()
+        self.notify_listener_thrad_pool = ThreadPoolExecutor(max_workers=1)
+
+    def do_initial_sync(self):
+        """Performs an initial sync of all metagraphs.
+
+        Unlike regular syncs, this will not notify listeners of the updated metagraph.
+        """
+        bt.logging.debug("Metagraph syncer do_initial_sync called")
+
+        for netuid in self.config.keys():
+            fn = functools.partial(self.subtensor.metagraph, netuid)
+            with self.lock:
+                metagraph = utils.run_in_subprocess(fn, ttl=120)
```

Comment:
do we need the lock around the run_in_subprocess or can we have it just around the state update? As you have it below. I suppose we are blocking on one metagraph update either way so just a minor consistency nit.

Reply:
Done